### PR TITLE
[8.x]: Rewrite Gradle build

### PR DESCRIPTION
**Pull request by lesserwhirls**
## Description of Changes

Rewrite the netCDF-Java gradle build from (mostly) scratch. The docs project build is too complex and will require more thought, but everything else has been rewritten using the kotlin DSL using Gradle 7.3.3 (upgrade from 6.9.1). While there are a lot of changes in this PR, it is probably most helpful to look over the new gradle doc ([gradle/README.md](https://github.com/lesserwhirls/netcdf-java/blob/wreckit/gradle/README.md)). The biggest workflow change is in the way tests are run (greatly simplified I think).

## PR Checklist

- [x] Indicate the version associated with this PR in the Title
       (e.g. "[5.x]: This is my PR title")
- [x] Link to any issues that the PR addresses
- [x] Add labels, especially if the PR should be ported to other versions
       (these labels start with "port: ")
- [x] Open as a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)
       until ready for review
- [x] Make sure GitHub tests pass
- [x] Mark PR as "Ready for Review"

### CODE REVIEW

**Comment by lesserwhirls** [comment]
wreck-it sean

**Comment by lesserwhirls** [comment]
It might be good to schedule a Google Meet to discuss these changes (with a demo).

File: README.md [hunk 1]
```diff
@@ -15,6 +15,14 @@ See branch [maint-5.x](https://github.com/unidata/netcdf-java/tree/maint-5.x).
 Version `4.6` is no longer supported outside of the context of the THREDDS Data Server (TDS).
 If you are looking for that codebase, it can be found at <https://github.com/Unidata/thredds/tree/4.6.x>.
 
+## Quick Start
+
+* build jar files: `./gradlew assemble`
```
Comment (JohnLCaron) [inline README.md:21]: add ./gradlew clean?

File: cdm-core/src/test/java/ucar/nc2/grid/TestGridDataset.java [hunk 2]
```diff
@@ -29,9 +29,9 @@ public void testBasics() throws Exception {
     try (GridDataset gds = GridDatasetFactory.openGridDataset(filename, errlog)) {
       assertThat(gds).isNotNull();
       System.out.println("readGridDataset: " + gds.getLocation());
-      assertThat(gds.toString()).startsWith("name = GFS_Puerto_Rico_191km_20090729_0000.nc\n"
-          + "location = ../cdm-core/src/test/data/ncml/fmrc/GFS_Puerto_Rico_191km_20090729_0000.nc\n"
-          + "featureType = GRID");
+      assertThat(gds.toString()).startsWith("name = GFS_Puerto_Rico_191km_20090729_0000.nc" + System.lineSeparator()
```
Comment (JohnLCaron) [inline cdm-core/src/test/java/ucar/nc2/grid/TestGridDataset.java]: thats painful. maybe change to use String.format(".... %n") ?

File: cdm-test/src/test/java/ucar/nc2/ncml/TestNcMLStrides.java [hunk 3]
```diff
@@ -69,8 +69,8 @@ private void testStride(String stride) throws IOException, InvalidRangeException
     Index ci = correct.getIndex();
     Index di = data.getIndex();
     for (int i = 0; i < data.getSize(); i++)
-      assert (data.get(di.set(i)) == correct.get(ci.set(i))) : stride + " index " + i + " = " + data.get(di.set(i))
-          + " != " + correct.get(ci.set(i));
+      assert (data.get(di.set(i)) == correct.get(ci.set(i)))
```
Comment (JohnLCaron) [inline cdm-test/src/test/java/ucar/nc2/ncml/TestNcMLStrides.java]: better to change to assertThat() whenever you have to touch this. But I left NcML tests out of that conversion, so I guess you can too.

File: gcdm/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    alias(libs.plugins.protobufPlugin)
+    alias(libs.plugins.execforkPlugin)
+}
+
+// Matches Maven's "project.description"
+description = "gRPC client and server implementation of CDM Remote Procedure Calls (gCDM)."
+
+dependencies {
+    implementation(platform(project(":netcdf-java-platform")))
+    implementation(platform(libs.grpcBom))
```
Comment (JohnLCaron) [inline gcdm/build.gradle.kts:23]: what is this?

File: netcdf-java-bom/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+plugins {
```
Comment (JohnLCaron) [inline netcdf-java-bom/build.gradle.kts:1]: whats diff between  netcdf-java-bom and  netcdf-java-platform ?

File: gcdm/build.gradle.kts (hunk 4 above)
Comment (JohnLCaron) [reply gcdm/build.gradle.kts:23]: what is this Bom thing?

**Comment by JohnLCaron** [comment]
Excellent work! Thanks so much

File: gcdm/build.gradle.kts (hunk 4 above)
Comment (lesserwhirls) [reply gcdm/build.gradle.kts:23]: The bom lists all of the grpc projects and their versions in one file:

https://repo1.maven.org/maven2/io/grpc/grpc-bom/1.43.1/grpc-bom-1.43.1.pom

Two nice things about using it. 1. You only specify the version once, and anything you reference from the bom picks up the version from the bom, and 2. using the bom as a gradle "platform" will also handle aligning the versions of transitive dependencies.

File: netcdf-java-bom/build.gradle.kts (hunk 5 above)
Comment (lesserwhirls) [reply netcdf-java-bom/build.gradle.kts:1]: The netcdf-java-bom lists the artifacts that make up the netcdf-java library, along with their version.

https://artifacts.unidata.ucar.edu/repository/unidata-snapshots/edu/ucar/netcdf-java-bom/8.0.0-SNAPSHOT/netcdf-java-bom-8.0.0-20211229.052136-33.pom

It gives users a one-stop-shop for accessing netcdf-java artifacts (like the grpc bom above).

The netcdf-java-platform lists all of the direct dependencies used by the artifacts listed in netcdf-java-bom. This is useful for something like the TDS, in which we want to make sure the versions of, say, protobuf, are aligned with what netCDF-Java uses.

https://artifacts.unidata.ucar.edu/repository/unidata-snapshots/edu/ucar/netcdf-java-platform/8.0.0-SNAPSHOT/netcdf-java-platform-8.0.0-20211229.052136-33.pom

I cleaned this up quite a bit, so the version in the link above includes some things that don't really matter to our users (like the bounce dependency that part of toolsUI).

Comment (lesserwhirls) [reply netcdf-java-bom/build.gradle.kts:1]: Also, we dogfood the netcdf-java-platform (you will see it at the top of the dependency list for every project), which ensures that each project uses the exact same versions of the core dependencies.

File: cdm-test/src/test/java/ucar/nc2/ncml/TestNcMLStrides.java (hunk 3 above)
Comment (lesserwhirls) [reply cdm-test/src/test/java/ucar/nc2/ncml/TestNcMLStrides.java]: Fixed :-)

File: cdm-core/src/test/java/ucar/nc2/grid/TestGridDataset.java (hunk 2 above)
Comment (lesserwhirls) [reply cdm-core/src/test/java/ucar/nc2/grid/TestGridDataset.java]: Yeah, that was fugly...sorry about that! Fixed!

File: README.md (hunk 1 above)
Comment (lesserwhirls) [reply README.md:21]: Good idea - added.

**Comment by lesserwhirls** [comment]
> Excellent work! Thanks so much

I think you will like the testing workflow much better.

`test` runs any test that isn't annotated with a category.

`extendedTests` runs anything tagged with one of the following categories:

~~~
NeedsCdmUnitTest
NeedsExternalResource
NotPullRequest
Slow
~~~

GitHub Actions: `./gradlew test`

Jenkins: `./gradlew test extendedTests`

The only additional requirement to run the `extendedTests` is the thredds-test-data datasets.

**Comment by lesserwhirls** [comment]
Should be good to go. Note I made a few changes to testing based on our conversations last week. See the Gradle [README](https://github.com/lesserwhirls/netcdf-java/blob/wreckit/gradle/README.md#testing) for more details.

**Review by haileyajohnson — APPROVED** [review]

**Pull request merged**